addon/components/course/objectives.js

```diff
@@ -0,0 +1,215 @@
+import { waitForProperty } from 'ember-concurrency';
+import { action } from '@ember/object';
+import scrollTo from 'ilios-common/utils/scroll-to';
+import { map } from 'rsvp';
+
+export default class CourseObjectivesComponent extends Component {
+  @service store;
+  @service intl;
+  @service flashMessages;
+
+  @tracked manageParentsObjective;
+  @tracked parentsBuffer = [];
+  @tracked manageDescriptorsObjective;
+  @tracked descriptorsBuffer = [];
+
+  @tracked newObjectiveEditorOn = false;
+  @tracked newObjectiveTitle;
+
+  @tracked objectives;
+  @tracked cohortObjectives;
+
+  @restartableTask
+  *load(event, [course]) {
+    if (!course) {
+      return;
+    }
+    this.objectives = yield (course.objectives).toArray();
+    this.cohortObjectives = yield this.getCohortObjectives(course);
+  }
+
+  get isManaging() {
+    return this.isManagingParents || this.isManagingDescriptors;
+  }
+
+  get isManagingParents() {
+    return !!this.manageParentsObjective;
+  }
+  get isManagingDescriptors() {
+    return !!this.manageDescriptorsObjective;
+  }
+
+  get showCollapsible(){
+    return this.objectives?.length && !this.isManaging;
+  }
+
+  async getCohortObjectives(course) {
+    const cohorts = (await course.cohorts).toArray();
+    const programYears = cohorts.map(c => c.belongsTo('programYear').id());
+    if (programYears.length) {
+      await this.store.query('objective', {
+        filters: {
+          programYears,
+        },
+      });
+    }
+    return await map(cohorts, async cohort => {
+      const programYear = await cohort.programYear;
+      const program = await programYear.program;
+      const objectives = (await programYear.objectives).toArray();
+      const objectiveObjects = await map(objectives, async objective => {
+        let competencyId = 0;
+        let competencyTitle = this.intl.t('general.noAssociatedCompetency');
+        const competency = await objective.competency;
+        if (competency) {
+          competencyId = competency.id;
+          competencyTitle = competency.title;
+        }
+        return {
+          id: objective.id,
+          title: objective.textTitle,
+          competencyId,
+          competencyTitle,
+          cohortId: cohort.id,
+        };
+      });
+      const competencies = objectiveObjects.reduce((set, obj) => {
+        let existing = set.findBy('id', obj.competencyId);
+        if (!existing) {
+          existing = {
+            id: obj.competencyId,
+            title: obj.competencyTitle,
+            objectives: []
+          };
+          set.push(existing);
+        }
+        existing.objectives.push(obj);
+        return set;
+      }, []);
+
+      return {
+        title: `${program.title} ${cohort.title}`,
+        id: cohort.id,
+        competencies,
+      };
+    });
+  }
+
+  @dropTask
+  *manageParents(objective) {
+    const cohortObjectives = yield waitForProperty(this, 'cohortObjectives', Array.isArray);
```

Comment:
this looks like it has the potential to lock up.
just thinking out loud here - would it be possible to init `cohortObjectives` as an empty array, and then check here (and elsewhere, template-side) if cohortObjectives not empty before allowing parent management?

Reply:
Yea, I have this on my list. To deactivate the button would require passing the status of this task down through several levels of component. I figured less change here is better and I'll refactor out `<CourseObjectiveList>` and '<CourseObjectiveListItem>` as a next step.

Reply:
fair enough.